vignettes/unit_test_guidance.Rmd

```diff
@@ -0,0 +1,151 @@
+frequently running the tests, they can assure their recent modifications to the 
+code haven’t broken anything. In other words, unit tests help prevent regressions.
+
+### Unit Tests Can Contribute to Higher Code Quality
+Since unit tests act as a safety net, developers become more confident when 
+changing the code. They can refactor the code without fear of breaking things, 
+driving the general quality of the code base up.
+
+### Unit Tests Can Contribute to Better Application Architecture
+If you can add unit tests easily to a code base, that’s usually a good sign 
+regarding the quality of the app’s architecture. So, the drive to write testable
+code can be an incentive for better architecture.
+
+### Detects Code Smells in your Codebase
+If ease of adding unit tests to a code base is a good sign, the opposite is also
+true. Having a hard time creating unit tests for a given piece of code might be 
+a sign of code smells in the code—e.g. functions that are too complex.
+
+## Writing Good Unit Tests
+
+### Tests Should Be Fast
+If they’re slow, developers won’t run them as often as they should. That defeats
+the whole purpose of having a suite of unit tests in the first place, which is 
+to boost the developers’ confidence to make changes to the code. The tests can’t
+work as the safety net if they’re not run often.
+
+### Tests Should Be Simple
+There are several techniques we can apply to have a high degree of confidence in
+the correctitude of our tests. One of those is to keep your tests with low 
+cyclomatic complexity. Cyclomatic complexity is a code metric that indicates the
+number of possible execution paths a given method can follow. A piece of code 
+with lower complexity is easier to understand and maintain, which means developers 
+are less likely to introduce bugs when working on it. We can measure the 
+cyclomatic complexity of your tests (using, for instance, a linter tool) and do 
+your best to keep it low.
+
+### Test Shouldn’t Duplicate Implementation Logic
+If the same person wrote both the test and the implementation, it’s possible they
+made the same errors in both places. Since the tests mirror the implementation,
+they might still pass, and the implementation could be wrong, but the tests might
+fool you into thinking otherwise. Resist the urge to make your tests fancy, keep
+them simple, and your testing suite will be better for it.
+
+### Tests Should Be Readable
+This best practice overlaps a little bit with the one about keeping your tests 
+simple. If tests are hard to read, developers are more likely to misunderstand 
+them and introduce bugs. Test cases could be used as a form of documentation, so
+they obviously need to be readable.
+
+### Running Unit Tests Part of the Build Process
+Automate the whole process of running the unit tests and taking some action when
+they fail. Your build process should execute your unit tests and mark the build 
+as broken when the tests fail.
+
+# Writing Unit Tests in Admiral
```

Comment:
Please use {admiral} throughout the vignette.

Reply:
Done.